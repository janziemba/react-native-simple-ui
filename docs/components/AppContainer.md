---
id: appcontainer
title: AppContainer
---
## Props
- `backgroundImage` ([_React Native Image's source_](https://facebook.github.io/react-native/docs/image#source), default null)
- `children` (_node_, **required**)

<img src="../assets/AppContainer/1.png" alt="AppContainer" width="250x" style="display: inline-block;">

## Examples

### Default
```jsx
<AppContainer>
    <Container>
        <Text>
            App container
        </Text>
    </Container>
</AppContainer>
```

### Background image
```jsx
<AppContainer backgroundImage={image}>
    <Container>
        <Text>
            App container with a background image
        </Text>
    </Container>
</AppContainer>
```
