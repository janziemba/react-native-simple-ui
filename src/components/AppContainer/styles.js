// @flow

import Color from 'color';

import { themeType } from '../../types';

export default ({ colors }: themeType): {} => {
    const imageBackgroundColor = Color(colors.layout.backgroundColor).alpha(0.96).rgb().toString();

    return {
        bottomSafeArea: {
            backgroundColor: colors.palette.primary,
            flex: 1,
        },
        content: {
            flex: 1,
        },
        imageBackground: {
            height: '100%',
            width: '100%',
        },
        imageBackgroundContent: {
            backgroundColor: imageBackgroundColor,
            flex: 1,
        },
        topSafeArea: {
            backgroundColor: colors.palette.primary,
            flex: 0,
        },
        wrapper: {
            backgroundColor: colors.layout.backgroundColor,
            flex: 1,
        },
    };
};
