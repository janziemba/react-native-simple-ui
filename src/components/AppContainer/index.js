// @flow

import Color from 'color';
import React, { PureComponent } from 'react';
import type { Node } from 'react';
import { ImageBackground, SafeAreaView, StatusBar, View } from 'react-native';

import withTheme from '../../themes/withTheme';
import styles from './styles';
import type { StylesType } from './styles';

type Props = {
    backgroundImage: Node,
    children: Node,
    styles: StylesType,
};

class AppContainer extends PureComponent<Props> {
    static defaultProps = {
        backgroundImage: null,
    };
    constructor(props: Props) {
        super(props);

        (this: any).renderContent = this.renderContent.bind(this);
    }
    renderContent() {
        const { backgroundImage, children, styles } = this.props;

        if (!backgroundImage) {
            return children;
        }

        return (
            <ImageBackground
                source={backgroundImage}
                style={styles.imageBackground}
            >
                <View style={styles.imageBackgroundContent}>
                    {children}
                </View>
            </ImageBackground>
        );
    }
    render() {
        const { styles } = this.props;

        const topSafeAreaIsLight: boolean = Color(styles.topSafeArea.backgroundColor).isLight();

        return (
            <View style={styles.wrapper}>
                <SafeAreaView style={styles.topSafeArea} />
                <SafeAreaView style={styles.bottomSafeArea}>
                    <StatusBar
                        barStyle={topSafeAreaIsLight ? 'dark-content' : 'light-content'}
                    />
                    <View style={styles.content}>
                        {this.renderContent()}
                    </View>
                </SafeAreaView>
            </View>
        );
    }
}

AppContainer = withTheme(styles, 'AppContainer')(AppContainer);

export default AppContainer;
