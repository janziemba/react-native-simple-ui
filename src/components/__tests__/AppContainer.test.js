import React from 'react';
import { View } from 'react-native';
import renderer from 'react-test-renderer';

import AppContainer from '../AppContainer';

it('renders correctly', () => {
    const tree = renderer.create(
        <AppContainer>
            <View />
        </AppContainer>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
