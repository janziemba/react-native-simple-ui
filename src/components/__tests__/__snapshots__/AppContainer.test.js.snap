// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      Object {
        "backgroundColor": "#158cba",
        "flex": 0,
      }
    }
  />
  <RCTSafeAreaView
    style={
      Object {
        "backgroundColor": "#158cba",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View />
    </View>
  </RCTSafeAreaView>
</View>
`;
